### ISSUE TITLE
1.5 - Make COMPOSE_PROJECT_NAME accessible to environment

### ISSUE THREAD

**Issue by mbontemps** [docker/compose#2294]
Given that: 
- [containers join the network under their container name](https://github.com/docker/compose/issues/2248) - at least for now
- container names depends on the project name

I think we need a quick basic solution for 1.5 to allow docker-compose with `--x-networking` enabled to **work on a project independently from the project name**. 

The reasoning: project names are variable (they can be different for two users depending on their directory, or can be changed in the CLI).  
So without forcing the project name, when we up a project, we have no way yet to predict the network name - and worse, no way to predict the hostnames of the other containers in the project.
## Proposed solution

I understand that nice solutions are planned for 1.6, but in the mean time I think we need a solution doable for 1.5. I suggest to set and make the environment variable `COMPOSE_PROJECT_NAME` available.

This way we can actually predict at which hostname other containers will be accessible in the current network. Something like `${COMPOSE_PROJECT_NAME}_redis_1`.
### What you do now

``` yaml
mysql:
    image: my-mysql

www:
    image: my-www
    environment:
        DB_HOST: hello_mysql_1
```

``` bash
~/Sites/hello$ docker-compose --x-networking up

~/Sites/hello$ docker-compose --x-networking --project-name staging up
# hello_mysql_1: Unknown host

~/Sites/dev$ docker-compose --x-networking up
# hello_mysql_1: Unknown host
```
### Proposal

The proposal is to make that possible:

``` yaml
mysql:
    image: my-mysql

www:
    image: my-www
    environment:
        DB_HOST: ${COMPOSE_PROJECT_NAME}_mysql_1
```

``` bash
~/Sites/hello$ docker-compose --x-networking up

~/Sites/hello$ docker-compose --x-networking --project-name staging up
# DB_HOST will be set to staging_mysql_1

~/Sites/dev$ docker-compose --x-networking up
# DB_HOST will be set to dev_mysql_1
```

**Comment by dnephin** [comment]
Good news, this works already!

https://github.com/docker/compose/blob/master/docs/reference/overview.md#compose_project_name
https://github.com/docker/compose/blob/master/docs/compose-file.md#variable-substitution

**Comment by mbontemps** [comment]
Thanks for the answer!

Exactly, it does *not* work by default… so you can not really rely on it.

In my case with:

```yaml
mysql:
    image: my-mysql

www:
    image: my-www
    environment:
        DB_HOST: ${COMPOSE_PROJECT_NAME}_mysql_1
```

The 1d and 3d usecase do not work:

```bash
~/Sites/hello$ docker-compose --x-networking up
```

The `COMPOSE_PROJECT_NAME` variable is not set so that does not work.

What do you think about setting this variable with the value which will actually be used by compose under the hood as (sanitized) project name?

**Comment by dnephin** [comment]
I'm hesitant to add something like that because it's likely only going to be necessary for 1 release. In the next release we should support aliases (so you won't need the project name), and we'd like to do #745 to persist the name.

It feels a bit like another hidden thing that may confuse users, although I can't think of a scenario right now where it would actually break something.

If it's a major issue, we could add it in a bug fix release.

**Comment by mbontemps** [comment]
Well IMO there are 2 things at play here:

1. This point is not actually related to networking. To be honest, after reading the doc I kinda expected `COMPOSE_PROJECT_NAME` to be accessible in all cases. Why wouldn't it be? Doesn't this feel more coherent than making it accessible *only if set manually in the CLI*? The current behavior seemed confusing to me until I read the source code. 

2. That being said, using this feature in networking to do what I want to do is indeed going to be useful only for 1 release, you're right.

But to me it is necessary as right now, I'm struggling to find *any* way to solve my problem in docker-compose with networking: running 2 times the same project with the same docker-compose on the same docker engine but with different networks. Think master + development branches for example.

So I would argue that given that it makes the environment variables behave more coherently and it does not seem to hurt, temporarily solving my scenario is just a bonus. :)

**Comment by pwaller** [comment]
I'm just hitting this exact situation. I have a configuration identical in spirit. I *really don't* want to have to configure anything except the project name, since our integration tests spin up many `compose up`'s at a time, and they need to not conflict. I don't want to have to teach my integration tests anything other than `docker compose up --project-name={unique string}`.

**Comment by dnephin** [comment]
`COMPOSE_PROJECT_NAME={unique string} docker-compose up` is not really all that much different from `docker-compose up --project-name={unique string}`.

**Comment by thaJeztah** [comment]
Also note this proposal in docker, that will allow setting aliases on containers so that they can be reached on the network, independent of the container name; https://github.com/docker/docker/issues/18699

**Comment by CircleCode** [comment]
> COMPOSE_PROJECT_NAME={unique string} docker-compose up is not really all that much different from docker-compose up --project-name={unique string}
in linux this is effectively not that much, but in windows, it is much more complicated.
Since one may have to build cross platforms compose files, this feature still remains blocking

**Comment by nottrobin** [comment]
It would be incredibly helpful to me to be able to somehow determine the project-name that `docker-compose` assigns to my folder by default. It's not simply the folder name, 'cos it removes, at least, hyphens.

Is there any simple way for me to get the auto-assigned project name, without overriding it manually?

My problem is that, after I do 

``` bash
docker-compose build web
````

I end up with a docker image called `{project-name}_web`. If I then want to clean up the system to leave it exactly as it was before, there's no way that I'm aware of to ask `docker-comose` to delete that built image. So I have to delete it manually with `docker rmi -f {project-name}_web`, but once again this is very difficult to script because I don't know what the {project-name} is.

**Comment by CircleCode** [comment]
@nottrobin wouldn't [environment file](https://docs.docker.com/compose/env-file/) solve your case?

**Comment by nottrobin** [comment]
@CircleCode How?

**Comment by CircleCode** [comment]
for example, you could add a `.env` file into your project folder, defining an explicit PROJECT_NAME
Then you do not need anymore to guess it since it is well known.

**Comment by nottrobin** [comment]
@CircleCode Thanks for responding.

I was always aware I could specify a project name, either through `COMPOSE_PROJECT_NAME` or through the `--project-name` option, but this is what I'm trying to avoid. `docker-compose` already has a mechanism for picking a project name for a given project, so it seems silly for me to implement another one.

Given that `docker-compose` is taking it upon itself to create a hopefully unique, readable project name by default, it seems entirely sensible to me that it would offer to tell you what that project name is - especially given that it does a less-than-perfect job of cleaning up after itself. Otherwise I feel like it should never have chosen one in the first place and instead ask you to specify one every time.

**Comment by CircleCode** [comment]
@nottrobin On the first glance, I agree with you.
Actually, I faced exactly the same problem in another way.

Why the decision has been took to automatically computing a project name is IMHO a bad question (and I do not even have an answer, I suppose this is a decision that made sense at the moment it has been decided). The fact is that it is now this way, and I found that the .env file is much better than `COMPOSE_PROJECT_NAME` or `--project-name` in the way it is explicit and persistent in the project since it is defined by a file of the project. For my needs, it was enough to rely on it rather than letting composer choose a project name.

Actually, I act as if
> \[compose\] never have chosen [a project name] in the first place and instead ask you to specify one every time

**Comment by nottrobin** [comment]
@CircleCode Yes all that makes sense. The problem I'm facing is that I'm currently trying to write tools for developers to run on their machines. The tools may be used on any project, and those projects may also be forked many different times (in different folders) on a developer's machine, and run simultaneously.

Therefore defining a single `COMPOSE_PROJECT_NAME` is out of the question, and even having one per project isn't great, because as I say I want to enable running a project multiple times in parallel. However, when developers run multiple versions of the project, they are almost always from directories with different names, and so the default `docker-compose` project naming mechanism works rather well.

But given everything you've said, it sounds like the only way forward for me is to re-implement a way of picking a project name which is unique enough for my needs, and always override the default.

**Comment by CircleCode** [comment]
@nottrobin 
Sorry if I wasn't able to help more.
BTW, remember I am in no way member of compose development team, nor affiliated to them, so maybe someone with better skills will be able to help you more.

**Comment by hainesc** [comment]
Please, I need it.

**Comment by lmakarov** [comment]
I have a similar use case and issue with this as @nottrobin  in https://github.com/docker/compose/issues/2294#issuecomment-224629014

In my case I need to connect a proxy container to any new project stack that is started on the host.
My current hacky solution to getting the compose project name is by stripping it out of the `docker-compose ps` output and relying on a particular pattern:

```bash
$ docker-compose ps cli
    Name                   Command               State        Ports       
-------------------------------------------------------------------------
projectname_cli_1   /opt/startup.sh gosu root  ...   Up      22/tcp, 9000/tcp 
```

```bash
$ network="$(docker-compose ps cli | awk 'NR==3 {print $1}' | sed 's/_cli_1//')_default"
$ echo $network
projectname_default
```

**Comment by r3m0t** [comment]
I would also like this looked at, my setup was:
```yaml
version: '2'
services:
  django:
    environment:
      - COMPOSE_PROJECT_NAME
```

In my django settings file, I added `'django.'+os.environ['COMPOSE_PROJECT_NAME']+'.docker'` to the `ALLOWED_HOSTS` as this matches the hostname assigned by dinghy-http-proxy. However, it turns out that environment only passes from the calling environment of `docker-compose`, rather than actually being set by `docker-compose` after it chooses a project name.

**Comment by emexelem** [comment]
Same issue here.
I have addressed the problem by declaring this alias:
`alias dc='COMPOSE_PROJECT_NAME=$(basename $(pwd)) docker-compose'`
With this, COMPOSE_PROJECT_NAME is available as environment variable in Docker compose files.
But I agree that's reimplementing a bit of what's in Docker, except that variable is not available as environment variable.

**Comment by wrouesnel** [comment]
I can't believe this is *still* an issue. I too would like this fixed, as it's relevant when you need to do things like docker inspect compose-created networks and volumes too.

**Comment by bersace** [comment]
`COMPOSE_PROJECT_NAME` is important to isolate code. We should have this information from the start.

For now, I use https://github.com/direnv/direnv to explicitly set `COMPOSE_PROJECT_NAME` like the shell alias above. But that's not portable.

**Comment by shin-** [comment]
@bersace Couldn't you use [a `.env` file](https://docs.docker.com/compose/env-file/)? Any way we could make that easier?

**Comment by bersace** [comment]
.env file is a nice idea, but it's not dynamic either.

**Comment by shin-** [comment]
I may be misunderstanding what you mean by that, but it's dynamic in the sense that you can override the default value in your `.env` file by setting an actual environment variable, or using the `-p` option in the command-line.

**Comment by bersace** [comment]
I avoid versionning .env and docker-compose.override.yml to let contributors customize the setup.

**Comment by oliveiraev** [comment]
+1 for me.

I'm another similar case. I run several applications that share the same docker-compose.yaml boilerplate. On my dev machine I can just let applications be called 'application', but, running on production, ```application``` string sometimes must be replaced by ```db.realappname.com``` or something similar.
Rely on ```COMPOSE_PROJECT_NAME``` is much more dynamical than updating a .env file.

I really don't know about something that would break allowing us to access ```COMPOSE_PROJECT_NAME``` everytime.

**Comment by jnovack** [comment]
Does the issue boil down to what @dnephin said??

> `COMPOSE_PROJECT_NAME={unique string} docker-compose up` is not really all that much different from `docker-compose up --project-name={unique string}.`

If so, I would expect them to work the same way.  Except they do not.

**Comment by oliveiraev** [comment]
Can someone on the team state use cases where this features is **NOT** desirable?

**Comment by hammy-lite** [comment]
We just hit this issue as well. The project name really should be accessible in the compose file no matter how it's specified, be it directory name, -p flag, or environment variable.

**Comment by SvenDowideit** [comment]
wow, really, one way of setting the project name sets an env var that is documented, and the other 2 ways (default dir name and `-p`) doesn't?  sounds at minimum something that should be documented, and considering that consistency simplifies both docs and UX, just simply made consistent!

/me goes off to explain to users that `docker-compose -p` is less useful than `COMPOSE_PROJECT_NAME=<blah> docker-compose` and should not be used.

**Comment by bersace** [comment]
A workaround would be to ask peers to have shell with always defined `COMPOSE_PROJECT_NAME`.

``` bash
$ cat autoproject.sh
cd() {
    builtin cd "$@"
    dirname=$(basename "${PWD,,}")
    export COMPOSE_PROJECT_NAME="${dirname//[ _.-]/}"
}
$ . autoproject.sh
$ cd .
$ echo $COMPOSE_PROJECT_NAME
testi
$ cd pqosudfq\ qdsf\ 89_D/
$ echo $COMPOSE_PROJECT_NAME
pqosudfqqdsf89d
```

That's intrusive and inefficient for open source projects.

**Comment by mw75** [comment]
As an alternative solution, i could imagine a "prepare-environment" hook on the top level of the configuration to allow any command to run to prepare the environment.

Also voted up the clean statement https://github.com/docker/compose/issues/2294#issuecomment-351262170

**Comment by JasonSwindle** [comment]
A silly work-around, if you are using systemd:

```
[Unit]
Description=%i service with docker compose
Requires=docker.service
After=docker.service

[Service]
Type=oneshot
RemainAfterExit=yes
WorkingDirectory=/etc/docker/compose/%i

# Set COMPOSE_PROJECT_NAME
# https://github.com/docker/compose/issues/2294
Environment=COMPOSE_PROJECT_NAME=%i

# Remove old containers, images and volumes
ExecStartPre=/usr/bin/docker-compose down --volumes --remove-orphans

# Compose up
ExecStart=/usr/bin/docker-compose up --detach --build

# Compose down, remove containers and volumes
ExecStop=/usr/bin/docker-compose down --volumes --remove-orphans

[Install]
WantedBy=multi-user.target
```

**Comment by yxliang01** [comment]
This has been here for a long time. Is the team planning to implement/won't fix this? I support this proposal.

**Comment by colinleroy** [comment]
Hi,
This merge request has been closed because one of the use cases of this issue has been resolved, but there are a lot more use cases where this would be useful (Bind mounts for example, etc). 
I think it should be solved just for consistency - what should I expect if I pass weird, conflicting values like `COMPOSE_PROJECT_NAME=project_A docker-compose -p project_B up` ?

**Comment by unclev** [comment]
I found this issue when I was going to create new issue "Make _default_ project name accessible in docker-compose.yml".

My use case is CI/CD when it publishes images tagged as branch name and creates directory per branch. For eg.,

```yml
version: '2.4'

services:

  backend-service1:
    image: private-registry/service1:${COMPOSE_PROJECT_NAME}
    environment:
      VIRTUAL_HOST: ${COMPOSE_PROJECT_NAME}.example.com
      VIRTUAL_PATH: /service1/

  backend-service2:
    image: private-registry/service2:${COMPOSE_PROJECT_NAME}
    environment:
      VIRTUAL_HOST: ${COMPOSE_PROJECT_NAME}.example.com
      VIRTUAL_PATH: /service2/
```
I'm aware about putting COMPOSE_PROJECT_NAME into `.env` file. But this would violate DRY principle, cluttering deploy scripts and instance content.

I know, changing environment affects only sub-processes, but (as a possible approach) docker-compose potentially can substituting calculated values of known `COMPOSE_` variables when doing interpolation.

**Comment by C-Duv** [comment]
> Does the issue boil down to what @dnephin said??
> 
> > `COMPOSE_PROJECT_NAME={unique string} docker-compose up` is not really all that much different from `docker-compose up --project-name={unique string}.`
> 
> If so, I would expect them to work the same way. **Except they do not.**

Exactly: using `--project-name` does not set a `COMPOSE_PROJECT_NAME` we can use in the `docker-compose.yml`, so both method does not behave the same way.

Without this, when using `--project-name`, we cannot do the kind of [naming @uncle suggested](https://github.com/docker/compose/issues/2294#issuecomment-503676806).

**Comment by verglor** [comment]
Please unstale - this issue still needs to be fixed.

**Comment by david-l-riley** [comment]
This has still not been resolved, five years later. Are PRs welcome?

**Comment by davidqmora** [comment]
It's marked stale because the devs have decided to ignore the requests. The proposed solution is not conducive to automation. The intuitive behavior would be for the -p switch to set the environment variable.

**Comment by hammy-lite** [comment]
Please do not mark as stale as this is feature we've missed in several projects without satisfying solution.
Asking all users to do annoying workarounds for something that should be a trivial fix is not very responsible.

**Comment by hendrics** [comment]
I think this should be available for consistency if nothing else, but also to work cross platform. We are using --project-name in ci and for some tests need to know the project name.

**Comment by jonatanrs** [comment]
Recent open issues on this topic.
> #8125
> #8143

**Comment by jonatanrs** [comment]
It would be good if someone responsible for this project could comment on this issue. (@thaJeztah)

**Comment by bersace** [comment]
not stale

**Comment by marioja** [comment]
Hello, I have a compose file that creates several services with several volumes.  I want to use the project name in the volume path.  I know that I can set COMPOSE_PROJECT_NAME before calling docker-compose but that is not what I want to do.  I want to use the compose project name that docker-compose derives from the environment, the -p arguments or its own default (directory basename) in the path for my volumes.  This is not possible because as a docker-compose config command will confirm, the COMPOSE_PROJECT_NAME environment variable is only define in each service and is not accessible in the docker-compose.yml.
Can this be implemented and should this be added to a new issue?

**Comment by 29axe** [comment]
Up

**Comment by hammy-lite** [comment]
> Please do not mark as stale as this is feature we've missed in several projects without satisfying solution. Asking all users to do annoying workarounds for something that should be a trivial fix is not very responsible.

No reasonable solution has been proposed yet so this is still an issue.

**Comment by C-Duv** [comment]
@stale, please un-stale this unresolved issue

**Comment by sergioisidoro** [comment]
Not stale
